pkg/datagatherer/versionchecker/versionchecker.go

```diff
@@ -0,0 +1,310 @@
+	"github.com/jetstack/preflight/pkg/datagatherer/k8s"
+)
+
+const (
+	// these are the keys used to look the file paths up from the supplied
+	// config
+	gcrTokenKey = "token"
+
+	acrUsernameKey     = "username"
+	acrPasswordKey     = "password"
+	acrRefreshTokenKey = "refresh_token"
+
+	ecrAccessKeyIdKey     = "access_key_id"
+	ecrSecretAccessKeyKey = "secret_access_key"
+	ecrSessionTokenKey    = "session_token"
+
+	dockerUsernameKey = "username"
+	dockerPasswordKey = "password"
+	dockerTokenKey    = "token"
+
+	quayTokenKey = "token"
+
+	selfhostedUsernameKey = "username"
+	selfhostedPasswordKey = "password"
+	selfhostedBearerKey   = "bearer"
+	selfhostedHostKey     = "host"
+)
+
+// Config is the configuration for a VersionChecker DataGatherer.
+type Config struct {
+	// the version checker dg will also gather pods and so has the same options
+	// as the dynamic datagatherer
+	Dynamic                     k8s.ConfigDynamic
+	VersionCheckerClientOptions vcclient.Options
+	// Currently unused, but keeping to allow future config of VersionChecker
+	VersionCheckerCheckerOptions vcapi.Options
+}
+
+// UnmarshalYAML unmarshals the ConfigDynamic resolving GroupVersionResource.
+func (c *Config) UnmarshalYAML(unmarshal func(interface{}) error) error {
+	aux := struct {
+		Dynamic struct {
+			KubeConfigPath    string   `yaml:"kubeconfig"`
+			ExcludeNamespaces []string `yaml:"exclude-namespaces"`
+			IncludeNamespaces []string `yaml:"include-namespaces"`
+		} `yaml:"k8s"`
+		Registries []struct {
+			Kind   string            `yaml:"kind"`
+			Params map[string]string `yaml:"params"`
+		} `yaml:"registries"`
+	}{}
+	err := unmarshal(&aux)
+	if err != nil {
+		return fmt.Errorf("failed to unmarshal version checker config: %s", err)
+	}
+
+	c.Dynamic.KubeConfigPath = aux.Dynamic.KubeConfigPath
+	c.Dynamic.ExcludeNamespaces = aux.Dynamic.ExcludeNamespaces
+	c.Dynamic.IncludeNamespaces = aux.Dynamic.IncludeNamespaces
+	// gvr must be pods for the version checker dg
+	c.Dynamic.GroupVersionResource.Group = ""
+	c.Dynamic.GroupVersionResource.Version = "v1"
+	c.Dynamic.GroupVersionResource.Resource = "pods"
+
+	c.VersionCheckerClientOptions.Selfhosted = map[string]*vcselfhosted.Options{}
+	registryKindCounts := map[string]int{}
+	for i, v := range aux.Registries {
+		registryKindCounts[v.Kind]++
+		switch v.Kind {
+		case "gcr":
+			data, err := loadKeysFromPaths([]string{gcrTokenKey}, v.Params)
+			if err != nil {
+				return fmt.Errorf("failed to load params for %s registry %d/%d: %s", v.Kind, i+1, len(aux.Registries), err)
+			}
+
+			c.VersionCheckerClientOptions.GCR.Token = data["token"]
+		case "acr":
+			data, err := loadKeysFromPaths([]string{acrUsernameKey, acrPasswordKey, acrRefreshTokenKey}, v.Params)
+			if err != nil {
+				return fmt.Errorf("failed to load params for %s registry %d/%d: %s", v.Kind, i+1, len(aux.Registries), err)
+			}
+
+			c.VersionCheckerClientOptions.ACR.Username = data["username"]
```

Comment:
this is super minor but since we have the consts should we just use them in the keys?

Reply:
Oops, good spot, fixed in 75a168cce1ed172014521ff1f58c84e123483f91